### PR TITLE
feat: add artworkResult root field for first-class error handling

### PR DESCRIPTION
**Pull request by mzikherman**
This is an alternate approach to https://github.com/artsy/metaphysics/pull/5523 , and tbh I like this one a lot better.

This elevates our `orError` pattern (from mutations), to actual fields - better allowing us to handle errors in a first-class way, vs. trying to use the basic `errors` key in the GraphQL response, or custom `extensions`.

This way, the artwork page/screen query can start to look like:

``` graphql
artwork(id: "...") {
  ... on Artwork {
    ...ArtworkApp_artwork
  }

  ... on ArtworkError {
    ...ArtworkErrorApp_artworkError
  }
}
```

with a partial artwork schema being available in the error case (now that https://github.com/artsy/gravity/pull/17283 is merged).

Looks like - 

Screen Shot 2024-01-26 at 2 04 51 PM

What do y'all think?

### CODE REVIEW

File: src/schema/v2/artwork/artworkOrError.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+const ArtworkErrorType = new GraphQLObjectType<any, ResolverContext>({
+  name: "ArtworkError",
+  description:
+    "An error type, potentially containing a partial artwork response",
+  fields: {
+    artwork: {
+      type: new GraphQLObjectType<any, ResolverContext>({
+        name: "PartialArtwork",
+        fields: {
```
Comment (mzikherman) [inline src/schema/v2/artwork/artworkOrError.ts:23]: These fields are as an example - imagine the error page showing the artist/artworks (with artist follow control), as well as a grid of related artworks (via `layers`).

We can flush them out better as we decide on what to show.

File: _schemaV2.graphql [hunk 2]
```diff
@@ -2692,6 +2692,16 @@ interface ArtworkEdgeInterface {
...
 
+# An error type, potentially containing a partial artwork response
+type ArtworkError {
+  artwork: PartialArtwork
+  errorDetail: ArtworkErrorDetail
+}
+
+type ArtworkErrorDetail {
+  statusCode: Int!
+}
```
Comment (damassi) [inline _schemaV2.graphql:2703]: If the idea here is to elevate and apply this more generally, could we make this a bit more system centric? Ie, something like 

```graphql
type PageError = ArtworkPageError | ShowPageError | FooPageError

type PageErrorDetail {
  statusCode: Int
}
```

Thinking these types are a bit too generic, and should this pattern spread there will soon be a lot of unnecessary duplication.

Comment (mzikherman) [reply _schemaV2.graphql:2703]: Yes, I also thought that the basic 'error' type here isn't artwork-specific and could be more general, but that the duplication wouldn't be _that_ bad  as it's a basic type (vs. duplicating more business-y schema concepts).

I think you're right and might as well introduce that abstraction to the error detail to start with.

I'll give that another pass.

File: src/schema/v2/schema.ts [hunk 3]
```diff
@@ -254,6 +255,7 @@ const rootFields = {
   artwork: Artwork,
   artworkAttributionClasses: ArtworkAttributionClasses,
   artworkMediums: ArtworkMediums,
+  artworkOrError: ArtworkOrError,
```
Comment (damassi) [inline src/schema/v2/schema.ts:258]: Ideally we don't have to add new fields, and our root types would be able to accomodate. For example:

```
artwork(id: "foo") {
  slug
}

artwork(id: "foo") {
  ... on Artwork {
    slug
  }
  
  ... on PageError {
    statusCode 
  }
}
```

**Review by damassi — COMMENTED** [review]
My first take here is that this seems nice, but would be better to flesh out some of the patterns a bit better and make it less verbose. There's certainly framework code we could work in here that, with a little bit of thinking, would make most of this automatic. 

That said, I tend to agree with @dzucconi that this feels a little like a solution in search of a problem. 

(Though I might feel a bit different if some more time were applied to the overall API design, so that an addition like this was transparent.)

File: src/schema/v2/schema.ts (hunk 3 above)
Comment (mzikherman) [reply src/schema/v2/schema.ts:258]: I would tend to agree, but that's a breaking change for clients. Switching to a union type like this requires a new field, so that's why I went with the `orError`.

Comment (damassi) [reply src/schema/v2/schema.ts:258]: This would be a breaking change? The union bits are additive:  

```
artwork(id: "foo") {
  slug 

  ... on Artwork {
    slug
  }
  
  ... on ArtworkError {
    slug 
  }
  
  ... on PageError {
    statusCode 
  }
}
```

**Comment by mzikherman** [comment]
> That said, I tend to agree with @dzucconi that this feels a little like a solution in search of a problem.

I don't fully agree - we tend to not handle errors particularly _well_ (hand-wave-y comment), and have used the first-class `orError` pattern for all mutations for quite a while. That tends to work quite well - having a 'nice' error schema that is first-class allowed for good feedback to users, field-level errors, and so forth.

This essentially elevates that concept to apply to query fields as well. It also lets us nicely handle unpublished artwork error cases (to start), which historically have been dead-ends and not friendly. While that's a small subset of total inventory, it's important we can handle those cases 'well' in a compelling way, and I believe the `orError` field approach sets that up - with it being totally opt-in.

> flesh out some of the patterns a bit better and make it less verbose. There's certainly framework code we could work in here that, with a little bit of thinking, would make most of this automatic

I don't want to get too carried away with framework-esque code or anything like that, yet. This is all still a bit early/spike-y, my goal is to have this working for artworks and have an improved artwork 404 page in Force.

I agree that the error schema can be fleshed out more, but I don't think we need to get into (yet) anything that's like - automatically creating an `orError` structure from an existing field w/ a helper wrapper. That kinda thing can come with more use-cases.

So other than the error schema, I'm not quite sure what patterns/framework code you may envision? I think w/ the better-structured error type, this could be mergeable.

File: src/schema/v2/schema.ts (hunk 3 above)
Comment (mzikherman) [reply src/schema/v2/schema.ts:258]: I kind of get what you're saying, but mechanically it is a breaking change unfortunately. If you switched `artwork` (like in your example) to be a union type, you can _only_ query it via fragments (`... on Artwork`) -> so all clients would break as they don't use that.

The only acceptable way to query it would be-

``` graphql
artwork(id: "foo") {
  ... on Artwork {
    slug
  }
  ...
}
```

so that's why we do need a new field, we can't really mix-in or decorate the existing.

**Comment by damassi** [comment]
The spike bit is showing here that a POC works, which has been accomplished, so that's good! 

It might be annoying to design something a little more general when a quick add and merge (via this PR) works for the 404 project, but new patterns kinda warrant it a bit. With this, we're adding new fields and types to an already bloated schema pretty unnecessarily (since we _can_ extend the existing root type, with a bit of work). 

Just suggesting that we don't rush this, as there's not exactly a pressing need. We can design something that upgrades. And make it a little friendlier for devs to implement elsewhere. 

Not blocking or anything! It just seems a little tech-debt-y, when it doesn't have to be with some more time spent on it.

File: src/schema/v2/schema.ts (hunk 3 above)
Comment (damassi) [reply src/schema/v2/schema.ts:258]: Can we really not do that? I can swear i've used fields like this in the past. But maybe i could be misremembering.

Comment (damassi) [reply src/schema/v2/schema.ts:258]: Can we really not do that? I can swear i've used fields like this in the past. But maybe i could be misremembering.

**Comment by mzikherman** [comment]
For the record I don't think polluting the root schema (which _is_ bloated) with new `orError`-style fields is actually a bad thing. They essentially supplant the basic version, and arguably we _should_ have had the `orError` style to start with all along! So I view it as paying down tech debt (but of course in the meantime, as with lots of long-term tech debt efforts, you temporarily have a more complex state - as with supporting both `artwork` _and_ `artworkOrError`).

File: src/schema/v2/schema.ts (hunk 3 above)
Comment (mzikherman) [reply src/schema/v2/schema.ts:258]: That's always been the case AFAIK. I view the `orError` pattern (while a 'touch' more verbose than regular `artwork`), as akin to preferring connections over lists. It's a touch more complex an object, but is more forward-looking and flexible in supporting other decorations and metadata that will come in the future.

That's why I'm sort of thinking we maybe could have had `artworkOrError` _instead of_ `artwork` all along as an optimal schema!

Comment (damassi) [reply src/schema/v2/schema.ts:258]: How about `error` being a field on `artwork`?

File: _schemaV2.graphql (hunk 2 above)
Comment (mzikherman) [reply _schemaV2.graphql:2699]: I rolled with `RequestError` for now.

File: src/schema/v2/schema.ts (hunk 3 above)
Comment (mzikherman) [reply src/schema/v2/schema.ts:258]: (also, `artwork` will error and return `"artwork": null` in the response, so you wouldn't be able to query for any top-level fields on `artwork`, like `error`)

Comment (MounirDhahri) [reply src/schema/v2/schema.ts:258]: @damassi it would have been awesome if that was possible. I also think it's a breaking change, unfortunately. 
I would as well vouch for a union type here instead of an error field.

**Comment by MounirDhahri** [comment]
@mzikherman I love this addition. With this we finally have a solution for something that we couldn't have done reliably before. I would only suggest though to be have a more verbose approach for the errors here instead of using the generic one. This can allow us to share potentially the error types across different interfaces in the future.

```graphql
artworkOrError(id: "...") {
# artworkResult(id: "...") { or whatever name we would like to come up with
  ... on Artwork {
    slug
  }

  ... on IsDeleted {
    message
  }

  ... on isUnavailable {
    reason
  }

}
```

We can then use something similar for Sales for example

```graphql
saleOrError(id: "...") {
  ... on Sale {
    slug
  }

  ... on IsDeleted {
    message
  }

  ... on isUnavailable {
    reason
  }

... onIsPrivate {
   message
  }

}
```

**Comment by mzikherman** [comment]
I pushed a commit to see how it might look in Force to use this pattern - https://github.com/artsy/force/commit/1c5574dd7ffad3d8d1c648454fa0d13ce0e45837. 

As expected it was pretty straightforward to port the current Force behavior (separate error page triggered via `throw new HttpError...`). What's not clear to me yet is how to render `UnpublishedArtworkApp` (vs. `ArtworkApp`), for instance - yet somehow propagate a 404 status code. That's a purely Force-side complexity though.

File: src/schema/v2/schema.ts (hunk 3 above)
Comment (araujobarret) [reply src/schema/v2/schema.ts:258]: I can't think of it in a way that won't break clients too. Union type is betters, is something we already do in the orders context in Exchange.

Comment (mzikherman) [reply src/schema/v2/schema.ts:258]: Going to resolve this convo since I _think_ we're in agreement - union types are great(!), and going with a new root field for this. Will be a fun trial!

**Comment by mzikherman** [comment]
I'd love to get this merged so that we can try to go ahead and use this in Force (and Eigen!). We can always walk this back if it doesn't get anywhere or we run into complexities.

## Current agreements

It seems like we're in agreement on rolling with a new root field which is a union type and so is forward-looking to represent some more nuanced 'states', vs. our current non-union field.

### Open question 1

One question is the naming - `artworkOrError` kind of matches our mutation response scheme, but isn't the most forward-looking for if we had other things into the union type besides 'errors'. I like @MounirDhahri 's suggestion of `artworkResult` in that case.

### Open question 2

The other question is about the error schema. For right now, the only thing we know we want to include is a status code - this is for web clients (Force) which need that. We also have a partial artwork response used to render a custom error page. Other fields like `details`, `message`, etc. - could possibly wait until they're needed.

In general, b/c the error schema isn't a core domain concept, if we need to iterate/duplicate that...I'm not personally super concerned. That feels relatively easily iterable. So if there was a type called `ArtworkError` or even more specific ones like `UnpublishedArtworkError`, and then we added `UnpublishedSaleError` later which is very similar...I think that's fine!

There's at least _some_ specificity required since for artworks we do want to return a partial artwork response, so `ArtworkError` will always need to exist. But, it can include a more generic `requestError: RequestError` field which would be shared amongst other 'result'-type union fields w/ error support. That's the case in this PR...and so I _think_ that can work for now? And we'll see how it goes.

### What next?

Basically I want to merge this! I will push an update that renames `artworkError` -> `artworkResult`, but no other changes.

Would you re-review (cc @MounirDhahri @damassi @dzucconi @araujobarret et al)? Otherwise I'll plan on clicking 'Merge' this week.

**Review by damassi — COMMENTED** [review]
I'm still up in the air about this and the addition of new root fields for these core types. 

Even if we could namespace things, like we do with viewer, then that would give people a place to go for "upgrading", without polluting the global scope with years of history behind it. I don't think this is necessary, though. 

For now this is just for the artwork page, but the aim here is clearly to introduce a new pattern of usage, one that will be largely generic; and as such we could use some compositional patterns for making this less verbose and manual. It would be great if a bit more time was spent trying to think through the possibilities. The template is: first class error handling for x type. Much of that work can be automated, especially with Javascript. 

And in any case, when doing

```graphql
{
  artwork(id: "foo") {
    slug
  }
}
```

we should be handling errors first class from here, even if it requires a nested field, or something else. It is not impossible, and resolvers are remarkably flexible. It's just easy to throw this out for something else, but at the expense of increasing the surface area of our graph. 

I mean, again, no blockers and I don't want to hold anything up, and maybe new root fields are fine, but our schema is already messy enough as it is, and as soon as this gets implemented in Eigen we're stuck with it.

(I understand that it can be tiresome to debate, and perhaps other folks have thoughts. No need to respond again to all of my points. I've said enough, but just want to leave with the encouragement that ideally, for whatever solution we come up with, it involves `artwork`, and modifying consumers of `artwork` from within such. If we can't "fix" `artwork` to do what we need then I would argue that we haven't tried hard enough.)

**Comment by mzikherman** [comment]
@damassi Thanks for your thoughts. As you point out - I think we're going to agree to disagree. And I won't quote + respond to points in your message even though I really want to, since....well, as you say - no need to rehash! You know my thoughts on it and we're sort of just reiterating our already-spoken thoughts at this point.

Nevertheless, thank you for the review and sharing your thoughts. See you on the other side!

**Review by joeyAghion — APPROVED** [review]
From my perspective, there are 2, similar, real-world use cases for this proposed way of requesting historical artwork data. That seems like a good enough reason to consider a new root type! If we want to keep our options open a little, we could front-load the Force changes (where we have more control), and have some flexibility to alter the design before Eigen adopts it.
